src/app/clusters/window-covering-server/window-covering-server.cpp

```diff
@@ -0,0 +1,601 @@
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+#include "window-covering-server.h"
+
+#include <app/util/af.h>
+
+#include <app/common/gen/attribute-id.h>
+#include <app/common/gen/cluster-id.h>
+#include <app/common/gen/command-id.h>
+
+#include <app/CommandHandler.h>
+#include <app/common/gen/attributes/Accessors.h>
+#include <app/reporting/reporting.h>
+#include <app/util/af-event.h>
+#include <app/util/af-types.h>
+#include <app/util/attribute-storage.h>
+#include <string.h>
+
+#ifdef EMBER_AF_PLUGIN_SCENES
+#include <app/clusters/scenes/scenes.h>
+#endif // EMBER_AF_PLUGIN_SCENES
+
+using namespace chip::app::Clusters;
+
+#define WC_PERCENT100THS_MAX 10000
+
+//------------------------------------------------------------------------------
+// WindowCover
+//------------------------------------------------------------------------------
+
+WindowCover WindowCover::sInstance;
+
+WindowCover & WindowCover::Instance()
+{
+    return sInstance;
+}
+
+WindowCover::WindowCover() {}
+
+void WindowCover::TypeSet(EmberAfWcType type)
+{
+    std::underlying_type<EmberAfWcType>::type value = static_cast<std::underlying_type<EmberAfWcType>::type>(type);
+    WindowCovering::Attributes::SetType(mEndPoint, value);
+}
+
+EmberAfWcType WindowCover::TypeGet(void)
+{
+    std::underlying_type<EmberAfWcType>::type value;
+    WindowCovering::Attributes::GetType(mEndPoint, &value);
+    return static_cast<EmberAfWcType>(value);
+}
+
+void WindowCover::ConfigStatusSet(WindowCover::ConfigStatus status)
+{
+    uint8_t value = (status.operational ? 0x01 : 0) | (status.online ? 0x02 : 0) | (status.liftIsReversed ? 0x04 : 0) |
+        (status.liftIsPA ? 0x08 : 0) | (status.tiltIsPA ? 0x10 : 0) | (status.liftIsEncoderControlled ? 0x20 : 0) |
+        (status.tiltIsEncoderControlled ? 0x40 : 0);
+    WindowCovering::Attributes::SetConfigStatus(mEndPoint, value);
+}
+
+const WindowCover::ConfigStatus WindowCover::ConfigStatusGet(void)
+{
+    uint8_t value = 0;
+    ConfigStatus status;
+
+    WindowCovering::Attributes::GetConfigStatus(mEndPoint, &value);
+    status.operational             = (value & 0x01) ? 1 : 0;
+    status.online                  = (value & 0x02) ? 1 : 0;
+    status.liftIsReversed          = (value & 0x04) ? 1 : 0;
+    status.liftIsPA                = (value & 0x08) ? 1 : 0;
+    status.tiltIsPA                = (value & 0x10) ? 1 : 0;
+    status.liftIsEncoderControlled = (value & 0x20) ? 1 : 0;
+    status.tiltIsEncoderControlled = (value & 0x40) ? 1 : 0;
+    return status;
+}
+
+void WindowCover::OperationalStatusSet(WindowCover::OperationalStatus status)
+{
+    uint8_t global = OperationalStateToValue(status.global);
+    uint8_t lift   = OperationalStateToValue(status.lift);
+    uint8_t tilt   = OperationalStateToValue(status.tilt);
+    uint8_t value  = (global & 0x03) | ((lift & 0x03) >> 2) | ((tilt & 0x03) >> 4);
```

Comment:
This does not make sense to me.  `(lift & 0x03) >> 2` and `(tilt & 0x03) >> 4` are always going to be 0, no?  Maybe this should be shifting before masking?  But then why does bitwise or on the results lead to something sensible?

Reply:
Boris's Lynx eyes:  arrows were inverted
FYI @bzbarsky-apple  this all code need to be remove in favor of bitfield 
Ricardo remove those because bitfield were not endian safe

Reply:
fixed

Reply:
@jmeg-sfy Fixed where?

Reply:
There arrows where indeed inverted. Fix pushed.